### ISSUE TITLE
Support Private Github repos via ssh keys (renamed)

### ISSUE THREAD

**Issue by MasterOdin** [Submitty/Submitty#2741]
To utilize private external resources, you either need to setup SSH keys or hard code a username and password into the VCS url. The former works well and is secure, but the latter will expose the username + password field to the students on their submission page, which is bad.

Like the VCS url, a username and password should be configurable at the system level and at the course level. 

If specified, for endpoints utilizing `://` scheme (http, ssh, etc), automatically insert the username + password into the appropriate spot, overwriting any existing username/password pair specified in the base url.

**Comment by MasterOdin** [comment]
In lieu of looking at #3305, I'm fully against actually implementing this and would almost suggest we prevent any users from using username/password pairs as there's several gotchas we have to be aware of including (depending on implementation), but not limited to:
* We don't leak the password on bash history
* We don't leak the password within the process list
* We don't leave the password in the .git/config folder

Our best bet would be doing something like https://serverfault.com/a/912788/367715 where we set the `GIT_ASKPASS` environment variable to point to python/bash script that then reads in the password from the DB/file, but it's potentially hacky and I'm not sure the risks it would have of leaking the password somewhere.

**Comment by bmcutler** [comment]
@MasterOdin 
once #3669 is merged, please go ahead and implement design for ssh keys or whatever you suggest

**Comment by MasterOdin** [comment]
SSH keys are already supported though and the documentation at https://submitty.org/sysadmin/git#using-external-git says how to do it. You'd just need to change https://github.com/Submitty/Submitty/pull/3669/files#diff-c7dfb660397d2883262efd7cdf1986a2R254 to use `git@github.com` instead of `https://github.com` for the clone line and then that would either clone (if public) or use the installed ssh keys. I suppose I could add a couple more words also on using `~/.ssh/config` file to specify any potential identify files one should use for Github.

**Comment by bmcutler** [comment]
Yeah, those instructions are now out of date.  
And there are no instructions for needing to create a special university/course github user.
And there are no instructions for students to grant access to that special user (or special users).

**Comment by bmcutler** [comment]
Can we query what github user ssh keys are already stored on the server?

UI for instructor:  would be nice to have specific name of github user here on this page, so instructor can write up details in their homework instructions.  If there are multiple ssh-key github users on the system, maybe they could radio button select the one they want students to use.

<img width="551" alt="Screen Shot 2019-05-10 at 3 24 48 PM" src="https://user-images.githubusercontent.com/5871417/57551744-d5ddf700-7337-11e9-9e5b-18052d47b2cf.png">

UI for student:  would be nice to paste the specific username here too.   And if there were multiple, just the one that the instructor selected.  

<img width="644" alt="Screen Shot 2019-05-10 at 3 22 53 PM" src="https://user-images.githubusercontent.com/5871417/57551627-85ff3000-7337-11e9-940b-0d27333c3817.png">

**Comment by MasterOdin** [comment]
I mean, we'd have to maintain (probably under a config file?) of the list of valid ssh keys and the associated github username. I'm not sure what the appropriate process for this would be. We could write a script that "provide github username/password" and then it sets up the ssh key, validates it, and then creates a new row in the config file for this user.

Or we could have the sysadmin generate the key and add it themselves, and then the script would just be responsible in looking at installed ssh keys for submitty_daemon (under /home/submitty_daemon/.ssh) and then asking which one should be associated with what github username (that they'd type in).

**Comment by bmcutler** [comment]
I'm in favor of having a submitty config file (that is manually maintained) that lists all available 'special' github users.  Generalized to (eventually) work for other git/vcs central repositories would be nice too :)

@MasterOdin Can you please take over implementation?  Thanks

**Comment by MasterOdin** [comment]
Closing this in-favor of #3672 and future other issues as this issue has become quite muddled with extraneous information on a direction we're not considering.
